---
comments: true
title: Single generation selection (SGS) test
date: '2020-06-17 12:00'
tags:
  - DelBay19
  - deltap
  - ouliter
  - null model
  - theta
  - drift
  - WGS
categories:
  - WGS data analysis
---

### Test for single-generation selection

In [Within-Generation Polygenic Selection Shapes Fitness-Related Traits across Environments in Juvenile Sea Bream](https://www.mdpi.com/2073-4425/11/4/398/htm#app1-genes-11-00398), the author estimated the expected distribution of allele frequency changes (deltap) due to finite sample sizes in the absence of selection to generate null expectations.

### Conceptional steps:

1) a panmictic common gene pool of finite size (i.e. the real population)

2) two samples of size N1 and N2 are drawn within the same generation

3) compare the observed allele frequency difference - ΔP=abs(P1−P2) with the distribution of ΔP expected from random sampling effects (i.e. due to finite sample size effects)

> Question: here the ΔP is estimated from absolute freqeuency changes. Shouldn't it to be the actual differences?

### Steps

1) 






[Experimental evidence for ecological selection on genome variation in the wild](https://onlinelibrary.wiley.com/doi/full/10.1111/ele.12238)

<img src="https://hzz0024.github.io/images/outlier/index.jpg" alt="img" width="800"/>

> Figure S2. Schematic of the null models of the absence of selection (null models 1 and 2). Each box or circle contains the genetic data for a hypothetical locus. Black numbers or letters denote individuals that survived and gray characters denote individuals that lived. A) In null model 1 the squares denote the true survivors and circles denote simulated sets of survivors. Density plots show the simulated distribution of allele frequency change for each locus and the vertical blue lines denote the observed allele frequency change. We reject the null hypothesis in model 1 for loci where the observed change (blue line) is an extreme tail of this null distribution. B) In null model 2 the real and replicate simulated data sets are show in boxes and numbers rather than genotypes are used to highlight the fact that the survival data is applied to all loci. Each of these data sets (the true data set and 100 replicate simulated data sets) are subjected to the null model 1 analysis (i.e., pane A) and the number of exceptional allele frequency change loci designated from the true survival data (vertical blue line) and replicate simulated data sets (gray histogram) are tabulated to determine whether the null hypothesis of no selection across the genome (null model 2) can be rejected.





   

