---
comments: true
title: DelBay19 outlier detection with survival vector method
date: '2020-08-14 12:00'
tags:
  - DelBay19
  - ouliter
  - null model
  - WGS
categories:
  - WGS data analysis
---

I put two plots here to illustrate the concept of this survival vector permutation method for outlier detection. 

<img src="https://hzz0024.github.io/images/pmt/design.jpg" alt="img" width="800"/>

<img src="https://hzz0024.github.io/images/pmt/format.jpg" alt="img" width="800"/>

So far I have finished the trial 1 and obtained ~ 17000 outliers. Here I used the quantile based approach to identify the potential outliers (see code below). This trial 1 assume the absence of seletion on individual loci.  

The whole process is seperated into three parts:

1) produce null data using a binomial survival vector, sampling 100 times.

2) estimate the allele frequency using the C++ based script published by [Kim et al 2011](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-12-231) (with minor modification). This is the script that angsd used for MAF estimation (i.e. -doMaf 1). I modified the script to ensure that the same allele is used for frequency esitmation. The allele frequency is obtained based on the genotype likelihoods, with the input as sequencing counts data (generated by -dumpcount 4 in the angsd).
   
3) for each SNP, calculate the deltap (from 100 sampling data) and compare the null distribution to the observed deltap

I look forward to the results from second trial.  
