---
comments: true
title: ngsLD
date: '2020-04-27 12:00'
tags:
  - CVreseq
  - DelBay19
  - LD
  - ngsLD
  - WGS
categories:
  - WGS data analysis
---

The ngsLD is a program that could estimate pairwise linkage disequilibrium (LD) taking the uncertainty of genotype's assignation into account. It would be interesting to compare the resequencing data (wild vs. a selected strain) and then also our low coverage data.

### Evaluate the SNP depth distribution

- DelBay19

```sh
angsd -b DelBay19_12.list -anc cv30.fa -ref cv30.fa -out depth/QC_DelBay19_12_maxD2000 \
-dosaf 1 -GL 1 -doGlf 2 -doMaf 1 -doPost 1 -doCounts 1 -doDepth 1 -P 20 -minInd 9 -minQ 20 -minMapQ 20 -minMaf 0.05 \
-remove_bads 1 -uniqueOnly 1 -only_proper_pairs 1 -SNP_pval 1e-6 -setMaxDepth 2000 -doMajorMinor 1 -rf chr.list >& QC_DelBay19_12_maxD2000.log
```
Total number of sites analyzed: 18595684     
Number of sites retained after filtering (minInd 6): 96849        
Number of sites retained after filtering (minInd 9): 51669    

<img src="https://hzz0024.github.io/images/ngsLD/QC_DelBay19_12_maxD2000.jpg" alt="img" width="800"/>

- CVreseq

```sh
angsd -b CVreseq_12.list -anc CVtot_3.0_newchr.fa.bgz -ref CVtot_3.0_newchr.fa.bgz -out depth/QC_CVreseq_12_maxD2000 \
-dosaf 1 -GL 1 -doGlf 2 -doMaf 1 -doPost 1 -doCounts 1 -doDepth 1 -P 20 -minInd 9 -minQ 20 -minMapQ 20 -minMaf 0.05 \
-remove_bads 1 -uniqueOnly 1 -only_proper_pairs 1 -SNP_pval 1e-6 -setMaxDepth 2000 -doMajorMinor 1 \
-rf chr.list >& QC_CVreseq_12_maxD2000.log
```
Total number of sites analyzed: 30608008    
Number of sites retained after filtering (minInd 6): 1614133
Number of sites retained after filtering (minInd 9): 1302240   

<img src="https://hzz0024.github.io/images/ngsLD/QC_CVreseq_12_maxD2000.jpg" alt="img" width="800"/>

---
### Generate the beagle file for ngsLD run

- DelBay19 
```sh
angsd -b DelBay19_12.list -anc cv30.fa -ref cv30.fa -out ngsLD/DelBay19_minI9D4maxD55_minQ20_minMAF05_SNPe6_chr10 \
-dosaf 1 -GL 1 -doGlf 2 -doMaf 1 -doPost 1 -doVcf 1 -doCounts 1 -doDepth 1 -dumpCounts 1 -P 20 -minQ 20 -minMapQ 20 -minMaf 0.05 \
-remove_bads 1 -uniqueOnly 1 -only_proper_pairs 1 -SNP_pval 1e-6 -setMinDepth 4 -setMaxDepth 55 -minInd 9 -doMajorMinor 1 \
-rf chr.list >& DelBay19_minI9D4maxD55_minQ20_minMAF05_SNPe6_chr10.log
```

- CVreseq
```sh
angsd -b CVreseq_12.list -anc cv30.fa -ref cv30.fa -out ngsLD/CVreseq_minI9D88maxD108_minQ20_minMAF05_SNPe6_chr10 \
-dosaf 1 -GL 1 -doGlf 2 -doMaf 1 -doPost 1 -doVcf 1 -doCounts 1 -doDepth 1 -dumpCounts 1 -P 20 -minQ 20 -minMapQ 20 -minMaf 0.05 \
-remove_bads 1 -uniqueOnly 1 -only_proper_pairs 1 -SNP_pval 1e-6 -setMinDepth 88 -setMaxDepth 108 -minInd 9 -doMajorMinor 1 \
-rf chr.list >& CVreseq_minI9D4maxD55_minQ20_minMAF05_SNPe6_chr10.log
```

### LD decay
```sh
Rscript --vanilla --slave fit_LDdecay.R --ld_files LD.list --out ref_ch.pdf --fit_level 3
```
Random seed: 65743
==> Fitting r2 LD decay assuming a three (rate of decay, max LD and min LD) parameter decay model   
              File LD   DecayRate     LDmax     LDmin     
1  ch_ngsLD_output r2 0.006095614 0.4581553 0.3406605      
2 ref_ngsLD_output r2 0.003160602 0.4473140 0.3243038      
==> Plotting data...     

---
### Conclusion and questions




