---
comments: true
title: ngsLD
date: '2020-04-27 12:00'
tags:
  - CVreseq
  - DelBay19
  - LD
  - ngsLD
  - WGS
categories:
  - WGS data analysis
---

The ngsLD is a program that could estimate pairwise linkage disequilibrium (LD) taking the uncertainty of genotype's assignation into account. It would be interesting to compare the resequencing data (wild vs. a selected strain) and then my low coverage data.

Hints that from ngsLD Github site:   
1) it may better to call SNPs based on MAF since monomorphic sites are not informative.     
2) ngsLD performance seems to drop considerable under extremely low coverages (<1x); consider these cases only if you have large sample sizes (>100 individuals).     
3) For some analyses (e.g. LD decay) consider sampling your data (--rnd_sample), since ngsLD will be much faster and you probably don't need all comparisons.            
4) For the LD decay, as a rule-of-thumb, consider using at least 10'000 SNPs; check the confidence interval and, if too wide, increase number of SNPs.

### Evaluate the SNP depth distribution

- DelBay19

```sh
angsd -b ch_6.bamlist -anc cv30.fa -ref cv30.fa -out depth/QC_ch_6_maxD2000 \
-dosaf 1 -GL 1 -doGlf 3 -doMaf 1 -doPost 1 -doCounts 1 -doDepth 1 -P 20 -minInd 4 -minQ 20 -minMapQ 30 -minMaf 0.05 \
-remove_bads 1 -uniqueOnly 1 -only_proper_pairs 1 -SNP_pval 1e-6 -setMaxDepth 2000 -doMajorMinor 1 -rf chr.list >& QC_ch_6_maxD2000.log
```
	-> Total number of sites analyzed: 11876980
	-> Number of sites retained after filtering: 19386 
  
<img src="https://hzz0024.github.io/images/ngsLD/QC_ch_6_maxD2000.jpg" alt="img" width="800"/>

```sh
angsd -b ref_6.bamlist -anc cv30.fa -ref cv30.fa -out depth/QC_ref_6_maxD2000 \
-dosaf 1 -GL 1 -doGlf 3 -doMaf 1 -doPost 1 -doCounts 1 -doDepth 1 -P 20 -minInd 4 -minQ 20 -minMapQ 30 -minMaf 0.05 \
-remove_bads 1 -uniqueOnly 1 -only_proper_pairs 1 -SNP_pval 1e-6 -setMaxDepth 2000 -doMajorMinor 1 -rf chr.list >& QC_ref_6_maxD2000.log
```
	-> Total number of sites analyzed: 13238610
	-> Number of sites retained after filtering: 26119 

<img src="https://hzz0024.github.io/images/ngsLD/QC_ref_6_maxD2000.jpg" alt="img" width="800"/>    

- CVreseq

```sh
angsd -b NEH.bamlist -anc CVtot_3.0_newchr.fa.bgz -ref CVtot_3.0_newchr.fa.bgz -out depth/QC_NEH_6_maxD2000 \
-dosaf 1 -GL 1 -doGlf 3 -doMaf 1 -doPost 1 -doCounts 1 -doDepth 1 -P 20 -minInd 4 -minQ 20 -minMapQ 30 -minMaf 0.05 \
-remove_bads 1 -uniqueOnly 1 -only_proper_pairs 1 -SNP_pval 1e-6 -setMaxDepth 2000 -doMajorMinor 1 \
-rf chr.list >& QC_NEH_6_maxD2000.log
```
	-> Total number of sites analyzed: 27373560
	-> Number of sites retained after filtering: 957677 

<img src="https://hzz0024.github.io/images/ngsLD/QC_NEH_6_maxD2000.jpg" alt="img" width="800"/>

```sh
angsd -b CS.bamlist -anc CVtot_3.0_newchr.fa.bgz -ref CVtot_3.0_newchr.fa.bgz -out depth/QC_CS_6_maxD2000 \
-dosaf 1 -GL 1 -doGlf 3 -doMaf 1 -doPost 1 -doCounts 1 -doDepth 1 -P 20 -minInd 4 -minQ 20 -minMapQ 30 -minMaf 0.05 \
-remove_bads 1 -uniqueOnly 1 -only_proper_pairs 1 -SNP_pval 1e-6 -setMaxDepth 2000 -doMajorMinor 1 \
-rf chr.list >& QC_CS_6_maxD2000.log
```
	-> Total number of sites analyzed: 28443753
	-> Number of sites retained after filtering: 1316679 

<img src="https://hzz0024.github.io/images/ngsLD/QC_CS_6_maxD2000.jpg" alt="img" width="800"/>

---
### Generate the glf file for ngsLD run

- DelBay19 
```sh

```

- CVreseq
```sh

```

---
### Conclusion and questions




