---
comments: true
title: DelBay adult batch effect tests
date: '2021-02-17 12:00'
tags:
  - DelBay 
  - WGS
  - Batch effect
categories:
  - WGS data analysis
--- 

Following some preliminary analyses, it looks like the PCA plots across all sequenced samples have shown some level of batch effect, as we sequenced two sample sets with different sequencers: NextSeq 2x150 for 2019 and Novoseq 2x150 for 2020 samples. On the other hand, we are also expecting to see some differentiation between the two years, possibly due to sweepstake reproduction and temporal fluctuations. Unfortunately, we did not include duplicate samples in the 2020 batch, and I can not tell if batch effect exists and/or to what extent the batch effect impacts the patterns. 

In this post I will do some diagnostic tests for batch effect following the suggestion by Nicolas. Below are the initial summary and PCA plots. 

#### data summary

| Data       | Mean | Deviation |  SD | Mean+3SD |
|------------|------|-----------|-----|----------|
| All_441    | 1222 |   93413   | 306 |   2139   |
| CH_all_205 | 681  |   42529   | 206 |   1300   |

All_441: dataset include all challenge (DelBay19 and 20, n=205) and wild (n=236) samples.                     
CH_all_205: dataset include all challenge (DelBay19 and 20, n=205) samples.    

#### PCA

- All_441 with inversions. Dataset include all challenge (DelBay19 and 20, n=205) and wild (n=236) samples.          

<img src="https://hzz0024.github.io/images/DelBay_adult/All_maf0.05_minq25_pctind0.7_CV30_masked.jpg" alt="img" width="800"/>

- CH_all_205. Dataset include all challenge (DelBay19 and 20, n=205) samples.

|                                          | CH_all_205 (withinvers) | CH_all_205 (noinvers) |
|------------------------------------------|-------------------------|-----------------------|
| Total number of sites analyzed           | 511494035               | 511494035             |
| Number of sites retained after filtering | 3839986                 | 3600542               |

- CH_all_205 with inversions. CH_all_205: dataset include all challenge (DelBay19 and 20, n=205) samples.

<img src="https://hzz0024.github.io/images/DelBay_adult/CH_all_maf0.05_minq25_pctind0.7_CV30_masked_withinvers.jpg" alt="img" width="800"/>

- CH_all_205 after removing inversions. CH_all_205: dataset include all challenge (DelBay19 and 20, n=205) samples.

<img src="https://hzz0024.github.io/images/DelBay_adult/CH_all_maf0.05_minq25_pctind0.7_CV30_masked_noinvers.jpg" alt="img" width="800"/>

1) Figures above show how inversions impact the population structure patterns among challenge samples. Same as the PCA plot using all (441) samples across different batches. The third plot (after removing inversions) is generated by removing three major inverions on Chr 1, 5 and 6. 

2) The reason why I am running these PCA analyses with combined datasets is to test whether batch effect exists in the dataset. Not unexpected, the batch effect resulting from the different sequencing methods and/or read depth seperates the individuals by years.

A few questions regarding following steps:

1) the CNV inversions identifed from high-coverage WGS study is composed of more than thousand regions, should I keep removing them? Yes.

2) What is the workaround for batch effect? Using all sample may bring in the batch effect, whereas creating site files seperatly may require common shared loci identification. I prefer the latter method by generating a site file from the DelBay19 samples (including both challenge and wild, given its lower mean depth), and then use this site file to call SNPs for DelBay20 datasets. As DelBay20 challenge samples have higher mean depth, it is expected that I can capture all identifed SNPs. After that, I am going to conduct combined Fisher's exact test among individual population.

### Workflow

According to some pre-analyses above, here is the workflow for following data analyese.

<img src="https://hzz0024.github.io/images/DelBay_adult/workflow.jpg" alt="img" width="800"/>

A couple of suggestion for diagnostic tests:

1) Evaluate average base composition at each read position using FastQC to detect polyG tails (which may survive polyG trimming and mapping)

<img src="https://hzz0024.github.io/images/batch_effect/DelBay19_base.jpg" alt="img" width="800"/>

<img src="https://hzz0024.github.io/images/batch_effect/DelBay20_base.jpg" alt="img" width="800"/>

Conclusion: the polyG issue is not that obvious to me, although for samples like SR0719r_014 has shown relative high %G at the end of reads. It has been trimmed by fastp.

2) Calculate the frequencies of different base substitutions in private alleles in each batch of data to detect signs of base calling bias.

3) Exclude private alleles of each batch of data from the PCA to verify that the PCA pattern is not an artifact.

4) Check if coverage differs between the two batches.