---
comments: true
title: Rerun DelBay19
date: '2021-01-22 12:00'
tags:
  - DelBay
  - Challenge
  - WGS
  - QC
categories:
  - WGS data analysis
--- 

The dataset in this post comes from the adult challenge experiment in summer, 2019. The reason I am re-analysing this dataset is due to haplotig-masked genome. An independent folder was created for this dataset - DelBay19. The major change from DelBay20 is the sample list and table. 

Before data QC, we need to create two files to provide the detailed information of sequenced samples.

sample_list.txt: sample prefix part before *_1.fastq.gz* or *_2.fastq.gz*
sample_table.txt: a tab deliminated table with the following six columns, strictly in this order:

  - `prefix` the prefix of raw fastq file names

  - `lane_number` lane number; each sequencing lane or batch should be
    assigned a unique identifier

  - `seq_id` sequence ID; this variable is only relevant when different
    libraries were prepared out of the same sample and were run in the
    same lane (e.g. if you wanted to include a replicate). In this case,
    seq\_id should be used to distinguish these separate libraries. If
    you only have a single library prepared from each of your samples
    (even if you sequence that same library across multiple lanes), you
    can just put 1 for all samples in this column.

  - `sample_id` sample ID; a unique identifier for each individual
    sequenced

  - `population` population name; the population or other relevant
    grouping variable that the individual belongs to

  - `data_type` data type; there can only be two possible entries: `pe`
    (for paired-end data) or `se` (for single end data). We need this in
    the table because for some of our processing steps, the commands are
    slightly different for paired-end and single-end data.

These files are located at            
[pool1](https://github.com/hzz0024/HG_Code_Bay/blob/master/DelBay19/sample_lists/fastq_list_1.txt)            
[pool2](https://github.com/hzz0024/HG_Code_Bay/blob/master/DelBay19/sample_lists/fastq_list_2.txt)             
[pool3](https://github.com/hzz0024/HG_Code_Bay/blob/master/DelBay19/sample_lists/fastq_list_3.txt)             
[pool4](https://github.com/hzz0024/HG_Code_Bay/blob/master/DelBay19/sample_lists/fastq_list_4.txt)             
[pool5](https://github.com/hzz0024/HG_Code_Bay/blob/master/DelBay19/sample_lists/fastq_list_5.txt)                     
[HG_Code_Bay/DelBay19/sample_list/fastq_table.txt](https://github.com/hzz0024/HG_Code_Bay/blob/master/DelBay19/sample_lists/fastq_table.txt)            

---

Steps below share the same script as DelBay20. I just put the locations here,

`QC and trimming` [https://github.com/hzz0024/HG_Code_Bay/tree/master/DelBay19/script/2_trim](https://github.com/hzz0024/HG_Code_Bay/tree/master/DelBay19/script/2_trim)            
`Mapping` [https://github.com/hzz0024/HG_Code_Bay/tree/master/DelBay19/script/4_map](https://github.com/hzz0024/HG_Code_Bay/tree/master/DelBay19/script/4_map)            









